Mathlib/Topology/UniformSpace/Dini.lean

```diff
@@ -0,0 +1,141 @@
+import Mathlib.Topology.UniformSpace.CompactConvergence
+
+/-! # Dini's Theorem
+
+This file proves Dini's theorem, which states that if `F n` is a monotone increasing sequence of
+continuous real-valued functions on a compact set `s` converging pointwise to a continuous function
+`f`, then `F n` converges uniformly to `f`.
+
+We generalize the codomain from `ℝ` to a normed latttice additive commutative group `G`.
+This theorem is true in a different generality as well: when `G` is a linearly ordered topological
+group with the order topology. This weakens the assumptions of the norm and commutativity, in
+exchange for strengthening to a linear order. This separate generality is not included in this file,
+but that generality was included in initial drafts of the original
+[PR #19068](https://github.com/leanprover-community/mathlib4/pull/19068) and can be recovered if
+necessary.
+
+-/
+
+open Filter Topology
+
+variable {ι α G : Type*} [Preorder ι] [IsDirected ι (· ≤ ·)] [Nonempty ι]
+variable [TopologicalSpace α] [NormedLatticeAddCommGroup G]
+
+section Unbundled
+
+open TopologicalAddGroup Metric
+
+variable {F : ι → α → G} {f : α → G}
+
+namespace Monotone
+
+/-- **Dini's theorem** If `F n` is a monotone increasing collection of continuous functions on a
+converging pointwise to a continuous function `f`, then `F n` converges locally uniformly to `f`. -/
+lemma tendstoLocallyUniformly_of_forall_tendsto
+    (hF_cont : ∀ i, Continuous (F i)) (hF_mono : Monotone F) (hf : Continuous f)
+    (h_tendsto : ∀ x, Tendsto (F · x) atTop (𝓝 (f x))) :
+    TendstoLocallyUniformly F f atTop := by
+  refine subsingleton_or_nontrivial G |>.elim (fun _ ↦ ?subsingleton) (fun _ ↦ ?_)
+  case subsingleton =>
+    rw [funext fun x => Subsingleton.elim (F x) f, TendstoLocallyUniformly]
+    -- missing `tendstoLocallyUniformly_const`
+    exact fun _ h x ↦ ⟨Set.univ, univ_mem, .of_forall fun _ _ _ ↦ mem_uniformity_of_eq h rfl⟩
+  have F_le_f (x : α) (n : ι) : F n x ≤ f x := by
+    refine _root_.ge_of_tendsto (h_tendsto x) ?_
+    filter_upwards [Ici_mem_atTop n] with m hnm
+    exact hF_mono hnm x
+  rw [tendstoLocallyUniformly_iff F f atTop UniformAddGroup.toUniformSpace_eq]
+  intro v hv x
+  obtain ⟨ε, ε_pos, hε⟩ := Metric.nhds_basis_ball.mem_iff.mp hv
+  simp_rw +singlePass [← tendsto_sub_nhds_zero_iff] at h_tendsto
+  obtain ⟨n, hn⟩ := h_tendsto x |>.basis_right nhds_basis_ball ε ε_pos |>.exists
+  refine ⟨_, (hF_cont n).sub hf |>.continuousAt.preimage_mem_nhds <| isOpen_ball.mem_nhds hn, ?_⟩
+  have mono_preimage_Ioo : Monotone (fun i => (F i - f) ⁻¹' Metric.ball 0 ε) := by
+    intro n m hnm x hx
+    simp only [Set.mem_preimage, Pi.sub_apply, mem_ball, dist_zero_right] at hx ⊢
+    refine norm_le_norm_of_abs_le_abs ?_ |>.trans_lt hx
+    simp only [abs_of_nonpos (sub_nonpos.mpr (F_le_f _ _)), neg_sub]
+    gcongr
+    exact hF_mono hnm x
+  filter_upwards [Ici_mem_atTop n] with m (hnm : n ≤ m) z hz
+  exact hε <| mono_preimage_Ioo hnm hz
```

Comment:
I played a bit with this (feel free to not fully adopt my suggestion if there's some bits you don't like).
```suggestion
  simp_rw [Metric.tendstoLocallyUniformly_iff, dist_eq_norm']
  intro ε ε_pos x
  simp_rw +singlePass [tendsto_iff_norm_sub_tendsto_zero] at h_tendsto
  obtain ⟨n, hn⟩ := (h_tendsto x).eventually (eventually_lt_nhds ε_pos) |>.exists
  refine ⟨{y | ‖F n y - f y‖ < ε}, ⟨isOpen_lt (by fun_prop) continuous_const |>.mem_nhds hn, ?_⟩⟩
  have mono : Monotone (fun i => {y | ‖F i y - f y‖ < ε}) := by
    intro n m hnm x hx
    refine norm_le_norm_of_abs_le_abs ?_ |>.trans_lt hx
    simp only [abs_of_nonpos (sub_nonpos_of_le (F_le_f _ _)), neg_sub, sub_le_sub_iff_left]
    exact hF_mono hnm x
  filter_upwards [eventually_ge_atTop n] with m hnm z hz
  exact mono hnm hz
```
The key change is to use more API from normed/metric spaces (in particular, no more `TopologicalAddGroup.tendstoLocallyUniformly_iff`). This allows you to get rid of the case disjunction on `G`. 

Then I also replaced some "continuity of `F n - f`" arguments by continuity of `fun y ↦ ‖F n y - f y‖` to turn balls into inequalities, but this is mostly a matter of taste/golfing.

Reply:
In fact maybe the lat `have` can just be inlined